core/utils/lock_free_queue.h

```diff
@@ -0,0 +1,112 @@
+#ifndef BESS_UTILS_LOCK_FREE_QUEUE_H_
+#define BESS_UTILS_LOCK_FREE_QUEUE_H_
+
+#include <glog/logging.h>
+#include <rte_malloc.h>
+
+#include "../kmod/llring.h"
+#include "./common.h"
+#include "./queue.h"
+
+namespace bess {
+namespace utils {
+
+/**
+ * A wrapper class for llring that extends the abstract class Queue. Takes a 
+ * template argument T which is the type to be enqueued and dequeued.
+ */
+template <typename T>
+class LockFreeQueue : public Queue<T> {
+ public:
+  static const size_t kDefaultRingSize = 256;
+
+  /**
+   * Construct a new queue. Takes the size of backing ring buffer (must power of
+   * two and entries available will be one less than specified. default is 256),
+   * boolean where if true, queue is in single producer mode or if false, queue
+   * is in multi producer mode, boolean where if true, queue is in single consumer
+   * mode, or if false, queue is in multi consumer mode. default for both booleans
+   * is true.
+   */
+  LockFreeQueue(size_t capacity = kDefaultRingSize, bool single_producer = true,
+                bool single_consumer = true)
+      : capacity_(capacity) {
```

Comment:
If capacity must be a power of two, you should check that here (with CHECK) and indicate that a non-power-of-two arg will die.